### PR TITLE
Add Flatpak support

### PR DESCRIPTION
**Pull request by TheEvilSkeleton**
Related to https://github.com/qarmin/czkawka/issues/15

I managed to make the GUI work, ~~but I don't think it deletes the files when I tell it to.~~ it works!

This is just an initial flatpak. I'll try to get it to work if others are willing to help.

If you want to test out the flatpak, make sure you have `flatpak-builder` installed, and run the following commands:

```bash
git clone https://github.com/TheEvilSkeleton/czkawka
cd czkawka/flatpak
flatpak-builder --force-clean --install-deps-from=flathub --install build-dir com.github.qarmin.Czkawka.yaml
```

---

P.S. if you have any suggestions, feel free to comment.

### CODE REVIEW

**Comment by qarmin** [comment]
Great work!

I tested it and looks that everything works - file deletion, previews, cache, saving/loading config etc.

I was able to install it but without --user flag and with sudo permissions `sudo flatpak-builder --force-clean --install-deps-from=flathub --install  build-dir com.github.qarmin.Czkawka.yaml` (Probably due outdated flatpak 1.6.5)

**Comment by TheEvilSkeleton** [comment]
That's fine. When you want to install a flatpak in your user, you need to add the repository in your user as well. So in this case, you need to install Flathub with `--user` in order to install it as the user.

That was my fault. I usually install flatpaks with `--user` so I copy pasted.

**Comment by TheEvilSkeleton** [comment]
@qarmin keep in mind that we made it easy to publish it in Flathub since the flatpak doesn't download during the build process (for redistribution purposes). I can help you with publishing it in Flathub if you want.

**Comment by qarmin** [comment]
> I can help you with publishing it in Flathub if you want.

It would be great.

I published now a new Czkawka version, because adds Alsa dependency needed by external library(Czkawka doesn't play any sounds just decode audio),

New version will require rebuild of cargo-sources.json and probably will require also adding some  `--socket=pulseaudio` or something things.

Sadly this commit with adding audio and zip dependencies broke Snap builds, so I think about reverting it.

I tried to test it, but I couldn't find any method to create json cargo sources file.

**Comment by TheEvilSkeleton** [comment]
> I published now a new Czkawka version, because adds Alsa dependency needed by external library(Czkawka doesn't play any sounds just decode audio),
> 
> New version will require rebuild of cargo-sources.json and probably will require also adding some `--socket=pulseaudio` or something things.

I updated its permissions and bumped it. Tell me how that goes.

> I tried to test it, but I couldn't find any method to create json cargo sources file.

Download the `flatpak-cargo-generator` Python file in the root of the source code of Czkawka at https://raw.githubusercontent.com/flatpak/flatpak-builder-tools/master/cargo/flatpak-cargo-generator.py, then run the command
```bash
python ./flatpak-cargo-generator.py ./Cargo.toml -o flatpak/cargo-sources.json
```

**Comment by qarmin** [comment]
Thanks, I checked `--socket=pulseaudio`option and it doesn't change anything so can be safely removed.

Looks that for now this Flatpak package provides more functionality than current Snap package which have missing support for external drives and some hidden folders in `/home/user`

**Comment by TheEvilSkeleton** [comment]
@qarmin by default, https://github.com/qarmin/czkawka/pull/203/commits/cd6a589f5ca93303874dc271cdb30d22480ea0d3 won't work since it asks for https://github.com/qarmin/czkawka/pull/203/commits/cd6a589f5ca93303874dc271cdb30d22480ea0d3#diff-55e4c74d40f6d322fcee02c419923278516387286a39e8daa8de64e0039a3328R29, which will only be available once this MR is merged. You can either comment that line or change the last two lines to:

```diff
-     url: https://github.com/qarmin/czkawka.git
-     tag: 2.3.0
+     url: https://github.com/TheEvilSkeleton/czkawka.git
```

**Comment by TheEvilSkeleton** [comment]
@qarmin feel free to merge it. I think it's good to go.

Also, I'm pretty sure this will be accepted in Flathub right off the bat.

File: flatpak/com.github.qarmin.Czkawka.metainfo.xml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    </p>
+  </description>
+  <launchable type="desktop-id">com.github.qarmin.Czkawka.desktop</launchable>
+  <screenshots>
+    <screenshot type="default">
+      <image>https://user-images.githubusercontent.com/41945903/104711404-9cbb7400-5721-11eb-904d-9677c189f7ab.gif</image>
+    </screenshot>
+  </screenshots>
+  <releases>
+    <release version="3.3.0" date="2021-01-15"/>
```
Comment (qarmin) [inline flatpak/com.github.qarmin.Czkawka.metainfo.xml]: 2.3.0 instead 3.3.0

**Review by qarmin — CHANGES_REQUESTED** [review]

File: flatpak/com.github.qarmin.Czkawka.yaml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+modules:
+- name: czkawka_gui
+  buildsystem: simple
+  build-commands:
+  - cargo --offline fetch --manifest-path Cargo.toml
+  - cargo --offline build --release
+  - install -Dm755 ./target/release/czkawka_gui -t /app/bin/
+  - install -Dm644 ./data/icons/com.github.qarmin.czkawka.svg -t /app/share/icons/hicolor/scalable/apps/
+  - install -Dm644 ./pkgs/com.github.qarmin.czkawka.desktop -t /app/share/applications/
+  - install -Dm644 ./flatpak/com.github.qarmin.Czkawka.metainfo.xml -t /app/share/metainfo
```
Comment (qarmin) [inline flatpak/com.github.qarmin.Czkawka.yaml:29]: I have error 
```
Running: install -Dm644 ./flatpak/com.github.qarmin.Czkawka.metainfo.xml -t /app/share/metainfo
install: cannot stat './flatpak/com.github.qarmin.Czkawka.metainfo.xml': No such file or directory
```

Comment (TheEvilSkeleton) [reply flatpak/com.github.qarmin.Czkawka.yaml:29]: https://github.com/qarmin/czkawka/pull/203#issuecomment-761699805. It cannot work until it's merged. You can change the URL to my Czkawka fork for testing.

Comment (qarmin) [reply flatpak/com.github.qarmin.Czkawka.yaml:29]: Oh, yeah, I forgot about it

**Review by qarmin — APPROVED** [review]

**Comment by qarmin** [comment]
Thanks!

**Comment by TheEvilSkeleton** [comment]
Keep in mind that this should be accepted in Flathub with minor changes needed at best. You just:
1. fork https://github.com/flathub/flathub/
2. change branch to `new-pr`
3. copy everything inside the `/flatpak` directory of this project, and paste them in the newly created branch
4. Attempt to submit an MR, read everything in the description check everything you can in the description, and complete the prerequisites if needed
5. Submit the MR

**Comment by TheEvilSkeleton** [comment]
Oh right... I completely forgot that since the tag is from `2.3.0`, which is a tag that did not include the `flatpak/` directory, it won't work. It should work on the next release.

**Pull request merged**